reconcile/utils/expiration.py

```diff
@@ -1,41 +1,41 @@
 import datetime
+from typing import Iterable, Optional, Protocol, TypeVar, Union, cast
 
+DATE_FORMAT = "%Y-%m-%d"
+
+
+class RoleProtocol(Protocol):
+    expiration_date: Optional[str]
 
-def has_valid_expiration_date(role: str) -> bool:
-    date_bool = True
-    if role is None:
-        return date_bool
-    else:
-        date_format = "%Y-%m-%d"
-        try:
-            date_bool = bool(datetime.datetime.strptime(role, date_format))
-        except ValueError:
-            date_bool = False
-        return date_bool
 
+DictsOrRoles = TypeVar(
+    "DictsOrRoles", bound=Union[Iterable[RoleProtocol], Iterable[dict]]
+)
 
-def role_still_valid(role: str) -> bool:
-    if role is None:
-        return True
-    else:
-        exp_date = datetime.datetime.strptime(role, "%Y-%m-%d").date()
-        current_date = datetime.datetime.utcnow().date()
-        if current_date < exp_date:
-            return True
-        return False
 
+def date_expired(role: str) -> bool:
+    exp_date = datetime.datetime.strptime(role, DATE_FORMAT).date()
+    current_date = datetime.datetime.utcnow().date()
+    return current_date >= exp_date
 
-def filter(roles: list[dict], key: str = "expirationDate") -> list[dict]:
+
+def filter(roles: Optional[DictsOrRoles]) -> DictsOrRoles:
     """Filters roles and returns the ones which are not yet expired."""
     filtered = []
-    for r in roles:
-        expiration_date = r[key]
-        if not has_valid_expiration_date(expiration_date):
+    for r in roles or []:
+        if isinstance(r, dict):
+            key = "expirationDate"
+            expiration_date = r.get(key)
+        else:
+            key = "expiration_date"
+            expiration_date = getattr(r, key, None)
```

Comment:
Why not `expiration_date = r.expiration_date`?

Reply:
That's a very good question! :)